src/auth/mysql.rs

```diff
@@ -0,0 +1,362 @@
+        let database_uri: String = String::from(
+            format!("mysql://{}:{}@{}:{}/{}", user, pass, host, port, database)
+        );
+        let authenticator = MySqlAuthenticator{database_uri};
+        match authenticator.test_connection() {
+            Ok(_) => Ok(authenticator),
+            Err(e) => Err(Error::GenericError(e.to_string())),
+        }
+    }
+
+    fn test_connection(&self) -> Result<&'static str, Error> {
+        let _ = self.connect()?;
+        Ok("Connection successful")
+    }
+
+    /// Connects to MySql Server
+    fn connect(&self) -> Result<MysqlConnection, Error> {
+        debug_!("Connecting to Mysql server");
+        let connection = MysqlConnection::establish(&self.database_uri)?;
+        Ok(connection)
+    }
+
+    /// Search for the specified user entry
+    fn search(
+        &self,
+        connection: &MysqlConnection,
+        search_user: &str
+    ) -> Result<Vec<User>, Error> {
+        use super::schema::users::dsl::*;
+        let results = users.filter(username.eq(search_user))
+            .load::<User>(connection)
+            .map_err(|e| Error::GenericError(
+                    format!("Database query failed: {}", e)
+            ))?;
+        Ok(results)
+    }
+
+    /// Hash a password with the salt. See struct level documentation for the algorithm used.
+    // TODO: Write an "example" tool to salt easily
+    pub fn hash_password(password: &str, salt: &[u8]) -> Result<String, Error> {
+        Ok(hex_dump(
+            Self::hash_password_digest(password, salt)?.as_ref(),
+        ))
+    }
+
+    /// Verify that some user with the provided password exists in the CSV database, and the password is correct.
+    /// Returns the payload to be included in a refresh token if successful
+    pub fn verify(
+        &self,
+        username: &str,
+        password: &str,
+        include_refresh_payload: bool,
+    ) -> Result<AuthenticationResult, Error> {
+        let user = {
+            let connection = self.connect()?;
+            let mut user = self.search(&connection, username).map_err(|_e| {
+                super::Error::AuthenticationFailure
+            })?;
+            if user.len() != 1 {
+                Err(super::Error::AuthenticationFailure)?;
+            }
+
+            user.pop().unwrap() // safe to unwrap
+        };
+        if username != user.username {
+            return Err(Error::Auth(super::Error::AuthenticationFailure));
+        }
+        let hash = test::from_hex(&user.pw_hash)?;
+        let salt = test::from_hex(&user.salt)?;
+
+        let actual_password_digest = Self::hash_password_digest(password, &salt)?;
+        if !verify_slices_are_equal(actual_password_digest.as_ref(), &*hash).is_ok() {
+            Err(Error::Auth(super::Error::AuthenticationFailure))
+        } else {
+            let refresh_payload = if include_refresh_payload {
```

Comment:
That is reasonable and my latest commit has made that change.

One thing to check is that rowdy is properly checking the signature when a refresh request comes in, otherwise a malicious user could escalate privileges or change which user they are acting as (since no password is supplied). 

At present, neither the LDAP connector or MySQL connector verifies that the user that is deserialised from the refresh token actually corresponds to an existing user in the database or directory, which is probably a good thing to check (but shouldn't be strictly necessary if the token signature is checked).